File: server/src/main/java/com/sellis/server/service/CategoryService.java
package com.sellis.server.service;

import com.sellis.server.dto.CategoryRequest;
import com.sellis.server.exception.BadRequestException;
import com.sellis.server.exception.ResourceNotFoundException;
import com.sellis.server.model.Category;
import com.sellis.server.repository.CategoryRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class CategoryService {

    private static final Logger log = LoggerFactory.getLogger(CategoryService.class);

    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    public List<Category> getAllActiveCategories() {
        return categoryRepository.findByActiveTrueOrderByDisplayOrderAsc();
    }

    public Category getCategoryById(String id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", "id", id));
    }

    public Category createCategory(CategoryRequest request) {
        if (categoryRepository.existsByNameIgnoreCase(request.getName())) {
            throw new BadRequestException("Category with name '" + request.getName() + "' already exists");
        }

        Category category = new Category();
        category.setName(request.getName());
        category.setSlug(toSlug(request.getName()));
        category.setDescription(request.getDescription());
        category.setDisplayOrder(request.getDisplayOrder());
        category.setActive(true);

        Category saved = categoryRepository.save(category);
        log.info("Created category: {}", saved.getName());
        return saved;
    }

    public Category updateCategory(String id, CategoryRequest request) {
        Category category = getCategoryById(id);

        categoryRepository.findByNameIgnoreCase(request.getName())
                .filter(existing -> !existing.getId().equals(id))
                .ifPresent(existing -> {
                    throw new BadRequestException("Category with name '" + request.getName() + "' already exists");
                });

        category.setName(request.getName());
        category.setSlug(toSlug(request.getName()));
        category.setDescription(request.getDescription());
        category.setDisplayOrder(request.getDisplayOrder());

        return categoryRepository.save(category);
    }

    public void deleteCategory(String id) {
        Category category = getCategoryById(id);
        category.setActive(false);
        categoryRepository.save(category);
        log.info("Soft-deleted category: {}", category.getName());
    }

    private String toSlug(String name) {
        return name.toLowerCase()
                .replaceAll("[^a-z0-9\\s-]", "")
                .replaceAll("\\s+", "-")
                .replaceAll("-+", "-")
                .trim();
    }
}

File: server-2/src/services/categoryService.js
const Category = require("../models/Category");
const {
  ResourceNotFoundException,
  BadRequestException,
} = require("../utils/errors");

function generateSlug(name) {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, "")
    .replace(/\s+/g, "-");
}

async function getAllActiveCategories() {
  return Category.find({ active: true }).sort({ displayOrder: 1 });
}

async function getCategoryById(id) {
  const category = await Category.findById(id);
  if (!category) {
    throw new ResourceNotFoundException("Category", "id", id);
  }
  return category;
}

async function createCategory({ name, description, displayOrder }) {
  const exists = await Category.findOne({
    name: { $regex: new RegExp(`^${name}$`, "i") },
  });
  if (exists) {
    throw new BadRequestException(`Category with name '${name}' already exists`);
  }

  const category = new Category({
    name,
    slug: generateSlug(name),
    description,
    displayOrder: displayOrder || 0,
    active: true,
  });

  const saved = await category.save();
  console.log(`Category created: ${saved.name}`);
  return saved;
}

async function updateCategory(id, { name, description, displayOrder }) {
  const category = await Category.findById(id);
  if (!category) {
    throw new ResourceNotFoundException("Category", "id", id);
  }

  if (name && name.toLowerCase() !== category.name.toLowerCase()) {
    const exists = await Category.findOne({
      name: { $regex: new RegExp(`^${name}$`, "i") },
      _id: { $ne: id },
    });
    if (exists) {
      throw new BadRequestException(
        `Category with name '${name}' already exists`
      );
    }
  }

  category.name = name || category.name;
  category.slug = generateSlug(category.name);
  category.description =
    description !== undefined ? description : category.description;
  category.displayOrder =
    displayOrder !== undefined ? displayOrder : category.displayOrder;

  return category.save();
}

async function deleteCategory(id) {
  const category = await Category.findById(id);
  if (!category) {
    throw new ResourceNotFoundException("Category", "id", id);
  }
  category.active = false;
  await category.save();
  console.log(`Category deleted (soft): ${category.name}`);
}

module.exports = {
  getAllActiveCategories,
  getCategoryById,
  createCategory,
  updateCategory,
  deleteCategory,
};
